src/boot/linux.c

```diff
@@ -119,16 +93,72 @@ EFI_STATUS linux_exec(
+
+        err = pe_kernel_check_no_relocation(kernel->iov_base);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        const PeSectionHeader *headers;
+        size_t n_headers;
+
+        /* Do we need to validate anyting here? the len? */
+        err = pe_section_table_from_base(kernel->iov_base, &headers, &n_headers);
         if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error loading kernel image: %m");
+                return log_error_status(err, "Cannot read sections: %m");
+
+        /* Do we need to ensure under 4gb address on x86? */
+        _cleanup_pages_ Pages loaded_kernel_pages = xmalloc_pages(
+                        AllocateAnyPages, EfiLoaderCode, EFI_SIZE_TO_PAGES(kernel_size_in_memory), 0);
+
+        uint8_t* loaded_kernel = PHYSICAL_ADDRESS_TO_POINTER(loaded_kernel_pages.addr);
+        FOREACH_ARRAY(h, headers, n_headers) {
+                if (h->PointerToRelocations != 0)
+                        return log_error_status(EFI_LOAD_ERROR, "Inner kernel image contains sections with relocations, which we do not support.");
+                if (h->SizeOfRawData == 0)
+                        continue;
+
+                if ((h->VirtualAddress < image_base)
+                    || (h->VirtualAddress - image_base + h->SizeOfRawData > kernel_size_in_memory))
+                        return log_error_status(EFI_LOAD_ERROR, "Section would write outside of memory");
+                memcpy(loaded_kernel + h->VirtualAddress - image_base,
+                       (const uint8_t*)kernel->iov_base + h->PointerToRawData,
+                       h->SizeOfRawData);
+                memzero(loaded_kernel + h->VirtualAddress + h->SizeOfRawData,
+                        h->VirtualSize - h->SizeOfRawData);
+        }
 
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image;
+        _cleanup_free_ EFI_LOADED_IMAGE_PROTOCOL* loaded_image = xnew(EFI_LOADED_IMAGE_PROTOCOL, 1);
+
+        EFI_LOADED_IMAGE_PROTOCOL* parent_loaded_image;
         err = BS->HandleProtocol(
-                        kernel_image, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), (void **) &loaded_image);
+                        parent, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), (void **) &parent_loaded_image);
         if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error getting kernel loaded image protocol: %m");
+                return log_error_status(err, "Cannot get parent loaded image: %m");
+
+        VENDOR_DEVICE_PATH device_node = {
+                     .Header = {
+                             .Type = MEDIA_DEVICE_PATH,
+                             .SubType = MEDIA_VENDOR_DP,
+                             .Length = sizeof(device_node),
+                     },
+                     .Guid = STUB_PAYLOAD_GUID,
+        };
+
+        _cleanup_free_ EFI_DEVICE_PATH* file_path = device_path_replace_node(parent_loaded_image->FilePath, NULL, &device_node.Header);
+
```

Comment:
so, it's lovely that we install a device path here now if we directly invoke the inner image. but this got me thinking: for the BS-based approach we could really do the same. we currently pass a NULL device path there...

(yeah, this doesn't really directly have anything to do with your PR, but it might be nice to make both paths work the same)